editoast/src/views/timetable/similar_trains/trains_traffic.rs

```diff
@@ -0,0 +1,307 @@
+    // Train by ID
+    trains_by_id: HashMap<usize, TrainTraffic>,
+}
+
+impl TrainsTraffic {
+    pub fn new(traffic_valid_from: Option<DateTime<Utc>>) -> Self {
+        let traffic_date = if let Some(date) = traffic_valid_from {
+            date
+        } else {
+            Utc::now()
+                .checked_sub_months(Months::new(6))
+                .expect("6 month ago date should be valid")
+        };
+        TrainsTraffic {
+            // Date from which traffic are valid (ie. can be imported)
+            traffic_valid_from: traffic_date,
+            // Trains by speedlimit
+            trains_by_speedlimit: HashMap::new(),
+            // Trains by rollingstock
+            trains_by_rollingstock: HashMap::new(),
+            // Trains by segment
+            trains_by_segment: HashMap::new(),
+            // Train by ID
+            trains_by_id: HashMap::new(),
+        }
+    }
+
+    pub fn len(&self) -> usize {
+        self.trains_by_id.len()
+    }
+
+    /// Load the traffic file
+    #[tracing::instrument(skip_all, level = "info", err, name = "Trains traffic loading")]
+    pub fn load_from_file(&mut self, file_path: PathBuf) -> anyhow::Result<()> {
+        // Open the file and read it
+        let file = File::open(file_path.as_path())?;
+        let reader = BufReader::new(file);
+
+        // Read each line from the file, we have a train
+        for (index, line_result) in reader.lines().enumerate() {
+            let line = line_result?;
+            if !line.is_empty() {
+                let json_train: JsonTrainTraffic = serde_json::from_str::<JsonTrainTraffic>(&line)
+                    .map_err(|err| {
+                        anyhow::anyhow!(
+                            "Failed to parse train traffic on line {}: {}",
+                            index + 1,
+                            err
+                        )
+                    })?;
+                let train_data = to_train_traffic(json_train, index);
+                self.add_train_traffic(train_data)
+                    .map_err(|err| anyhow::anyhow!("Failed to add train traffic: {}", err))?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Retrierve a train by its id
+    pub fn get_by_id(&self, id: usize) -> Option<TrainTraffic> {
+        self.trains_by_id.get(&id).cloned()
+    }
+
+    /// Add a train traffic
+    pub fn add_train_traffic(&mut self, train_data: TrainTraffic) -> anyhow::Result<()> {
+        // Only import train that have a valid start date
+        if train_data.start_time > self.traffic_valid_from {
+            // Index the train by its id
+            self.trains_by_id.insert(train_data.id, train_data.clone());
+
+            // Index the train by its speedlimit
+            let trains_for_speedlimit = self
+                .trains_by_speedlimit
+                .entry(train_data.speed_limit_tag)
+                .or_default();
+            trains_for_speedlimit.insert(train_data.id);
```

Comment:
nitpick: You should be able to inline all of it.


```suggestion
            let trains_for_speedlimit = self
                .trains_by_speedlimit
                .entry(train_data.speed_limit_tag)
                .or_default()
                .insert(train_data.id);
```